공연홀 정보 추가 기능 구현

tlsgkr7416 opened the pull request:
#28 

- [x] 공연홀 정보 추가 기능 구현
- [x] 동일한 공연홀 중복 예외 구현
- [x] 공연홀 추가시 트랜잭션 처리

tlsgkr7416 committed:
Merge remote-tracking branch 'origin/feature/25/jh' into feature/28/sh

tlsgkr7416 committed:
공연홀 정보 추가 기능 구현

공연홀 정보(공연홀Id(pk), 공연장Id(fk), 이름, 좌석(열), 좌석(행), 좌석수)
공연홀 요청: venueHall
공연홀 controller: VenueController
공연홀 service: VenueHallService
공연홀 mapper: VenueHallMapper
공연홀 예외처리: VenueHallDuplicationException (공연홀은 list 형식으로 넣어 주기 때문에 이름이 중복 되면 예외처리)
기능: 공연홀 list 형식으로 넣어주기, checkDuplicationVenueHallName메소드로 공연홀 이름 중복 검사

tlsgkr7416 committed:
venueHall테이블 추가

etc에 venueHall 테이블 추가

tlsgkr7416 committed:
같은 공연장에서 이미 존재하는 공연홀을 추가 시 예외 처리

같은 공연장에 공연홀을 더 추가하고 싶을 때 이미 존재하는 공연홀이 있을 수 있으므로 이때 예외 처리를 해줌
VenueController의 checkVenueHallExists 메서드에서 이미 존재하는 공연홀이 있을 때 VenueHallAlreadyExistsException로 예외 처리
VenueHallMapper에서 <select id=isVenueHallExists...> 같은 경우 EXISTS(...)로 boolean을 리턴 시켰고
foreach 같은 경우 공연홀이 list 형식으로 여러 개 받아 올 수 있으므로 추가해 줬다.
IN은 공연홀을 추가한 name 중 이미 존재하는 name이 있다면 select를 시키도록 했다.
그리고 AND venueId = #{venueId}는 같은 공연장에 있다는 조건도 추가시켜줬다.

hanwix2 commented on src/main/java/com/show/showticketingservice/exception/venueHall/VenueHallDuplicationException.java:
```diff
@@ -0,0 +1,8 @@
+    public VenueHallDuplicationException() {
+        super("중복되는 공연홀이 존재합니다.");
+    }
+}
```
관련 로직을 보지 않고 얼핏 봤을 때 `VenueHallAlreadyExistsException ` 예외와 의미상 겹치는 것처럼 보이는데
나눠서 처리해야 한다면 예외명이나 메시지를 혼동되지 않도록 적용하는 건 어떨까요?

hanwix2 commented on src/main/java/com/show/showticketingservice/service/VenueHallService.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    public void checkDuplicationVenueHallName(List<VenueHall> venueHalls) {
+        Map<String, Integer> hashMap = new HashMap<>();
```
객체명을 단순히 hashMap 이 아니라 명확한 활용을 알아 볼 수 있는 이름은 어떨까요?  예를 들자면 hallNameMap 같은..
여기는 짧은 코드라 금방 이해가 되긴하지만 만약 로직이 길었다면 혼동이 있을 것 같습니다.

hanwix2 commented on src/main/resources/mybatis/mappers/VenueHallMapper.xml:
```diff
@@ -0,0 +1,29 @@
+
+<mapper namespace="com.show.showticketingservice.mapper.VenueHallMapper">
+
+    <insert id="insertVenueHall" parameterType="map">
+        INSERT INTO venueHall(venueId, name, rowSeats, columSeats, seatingCapacity)
+           VALUES
+              <foreach collection="venueHalls" item="venueHall" separator=",">
+                  (
+                  #{venueId},
+                  #{venueHall.name},
+                  #{venueHall.rowSeats},
+                  #{venueHall.columSeats},
+                  #{venueHall.seatingCapacity}
+                  )
+              </foreach>
+    </insert>
```
입력한 '좌석수'가 '행 수'와 '열 수'를 곱한 값과 다르다면 추후 서비스를 제공할 때 문제가 발생하지 않을까요?

추가적인 확인 과정을 거치거나, 
행과 열 값만 입력받고 계산한 값을 총 좌석수로 저장하는 방법은 어떤가요?

f-lab-dev commented on src/main/java/com/show/showticketingservice/controller/VenueController.java:
```diff
@@ -0,0 +1,32 @@
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/venue")
```
여기는 복수형으로 표기해주시는게 좋을 것 같네요

f-lab-dev reviewed (commented):
이 브랜치를 봤을때 정한님의 작업내용과 겹쳐보이는게 있는데 컨플릭트가 날 것 같습니다~
`git merge` 명령어로 정한님의 작업내용을 가져온 다음, 거기에 이어서 작업하는게 좋았었을 것 같아요~

tlsgkr7416 committed:
Merge remote-tracking branch 'origin/feature/25/jh' into feature/28/sh

tlsgkr7416 committed:
공연홀 추가 기능 피드백 반영

좌석(행) 좌석(열)을 생성자 내부에서 곱해서 총 좌석 수를 구함
venueHall 테이블 좌석(행, 열) int로 수정
공연홀 예외처리 클래스 수정

tlsgkr7416 committed:
공연홀 추가 메소드에 트랜잭션 구현

tlsgkr7416 committed:
Merge remote-tracking branch 'origin/feature/25/jh' into feature/28/sh

tlsgkr7416 committed:
mppper 메소드명 수정

isVenueHallExists => isVenueHallsExists 수정

tlsgkr7416 committed:
공연홀 정보 수정 Controller, Service 구현 (미완성)

controller updateVenueHall 메서드 선언
service checkVenueHallExists 오버로딩

tlsgkr7416 committed:
Merge branch 'feature/28/sh' into feature/29/sh

tlsgkr7416 committed:
공연 홀 정보 수정 기능 구현

controller: updateVenueHall구현
service: checkVenueHallExists 메소드로 공연 홀 이름이 이미 존재하는지 체크 updateVenueHall
메소드로 공연 수정 업데이트
mapper: updateVenueHall 공연 홀 수정, isVenueHallExists 수정한 공연 홀 이름이 이미 존재하나 체크

tlsgkr7416 committed:
공연 홀 service, mapper 메소드명 수정

tlsgkr7416 committed:
merge 충돌 해결

f-lab-dev commented:
나머지 리뷰들도 계속 적용해주세요~

tlsgkr7416 committed:
VenueHall validation수정

tlsgkr7416 committed:
공연 홀 수정 기능 트랜잭션 처리

tlsgkr7416 committed:
공연 홀 삭제 기능 구현

controller: VenueController의 deleteVenueHall을 통해 삭제 핸들러 구현
한번에 공연 홀을 여러개 삭제 할 수 있도록 클라이언트로부터 list 형식을 받을 수 있게 했다.
service: VenueHallService
mapper: VenueHallMapper

tlsgkr7416 committed:
공연 홀 정보 조회 기능 구현

controller: VenueController
service: VenueHallService
mapper: VenueHallMapper
공연장 id를 통해 공연 홀 정보를 db에서 받아와 list 형식으로 클라이언트에게 return 시켜줌

tlsgkr7416 committed:
오타 수정, 클래스명 수정

tlsgkr7416 committed:
메소드명 수정

tlsgkr7416 committed:
insertVenueHalls 메소드명 수정

tlsgkr7416 committed:
mapper select문 명시적으로 수정

tlsgkr7416 committed:
필요 없는 import문 제거

tlsgkr7416 committed:
Merge pull request #47 from f-lab-edu/feature/46/sh

[#46] 공연 홀 정보 조회 기능 구현

tlsgkr7416 committed:
Merge pull request #44 from f-lab-edu/feature/30/sh

공연 홀 삭제 기능 구현

tlsgkr7416 committed:
merge 충돌 수정

tlsgkr7416 committed:
merge 충돌 수정
